Fix String.scan to return NONE

yonta opened the pull request:
In [basis](http://sml-family.org/Basis/string.html#SIG:STRING.scan:VAL) description, `String.scan` retuns `NONE` if first character is non-printable.
But, `String.scan` in MLton returns `SOME ([], restStream)` now.

This PR is adding checking whether return value is nil or not.

## Example Code

```sml
val () =
    let
      (* scan starting '\n' that is not printable character *)
      val result = String.scan Substring.getc (Substring.full "\n")
    in
      case result of
          NONE => print "NONE\n"
        | SOME (str, _) => print ("SOME:" ^ str ^ "\n")
    end
```

## Expected

```
NONE
```

Actual

```
SOME:
```

MatthewFluet commented:
Thanks!  Agreed that MLton is deviating from the Basis Library spec for `String.scan`.

We should check the interaction of `String.scan` and `String.fromString`; some of the regression tests are now failing; see https://travis-ci.org/MLton/mlton/jobs/658684677#L2430.  It may be that the regression tests are incorrect (but, if so, then they should be updated as part of this PR).

yonta commented:
I see.
I check detail of regression test which is failing, now. :)

yonta commented:
I did it. 👍 
The Basis Library spec of `String.scan` is more complicated than I thought. 🤔 

- Updated a test20 in string test.
    - changed expected values to `NONE` from `SOME value`.
    - removed one case that `String.scan` should returns `SOME ([], _)`.
- Updated expected values of string.fromString test.
    - expected values is brought from conversion examples of `String.fromString` in Basis Library.
- Updated `String.scan` to follow Basis Library spec.
- Added `Char.formatSequenceOpt` function in `CHAR_EXTRA` module, to check whether input starts with escaped formatting characters.

MatthewFluet commented:
Looks good.  Thanks!
